File: FresherAndroid/AndroidBasicCode/ManhPT14_IVI_BAD_Practice_T01/app/src/main/java/com/example/contactapp/adapter/ContactAdapter.java
package com.example.contactapp.adapter;

import android.graphics.Bitmap;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.contactapp.R;
import com.example.contactapp.model.Contact;
import com.example.contactapp.utils.ImageUtils;

import java.util.ArrayList;
import java.util.List;

public class ContactAdapter extends RecyclerView.Adapter<ContactAdapter.ViewHolder> {

    public interface OnItemClickListener {
        void onItemClick(Contact contact);
    }

    private List<Contact> contacts = new ArrayList<>();
    private final OnItemClickListener listener;

    public ContactAdapter(OnItemClickListener listener) {
        this.listener = listener;
    }

    // Cập nhật danh sách (gọi từ Fragment sau khi load DB)
    public void setContacts(List<Contact> list) {
        this.contacts = list != null ? list : new ArrayList<>();
        notifyDataSetChanged();
    }

    public List<Contact> getContacts() {
        return contacts;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_contact, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Contact contact = contacts.get(position);
        holder.bind(contact, listener);
    }

    @Override
    public int getItemCount() {
        return contacts.size();
    }

    // ── ViewHolder ──────────────────────────────────────────────────────────
    static class ViewHolder extends RecyclerView.ViewHolder {
        private final TextView  tvName;
        private final TextView  tvPhone;
        private final ImageView imgAvatar;
        private final TextView  tvAvatarLetter;

        ViewHolder(@NonNull View itemView) {
            super(itemView);
            tvName         = itemView.findViewById(R.id.tvContactName);
            tvPhone        = itemView.findViewById(R.id.tvContactPhone);
            imgAvatar      = itemView.findViewById(R.id.imgAvatar);
            tvAvatarLetter = itemView.findViewById(R.id.tvAvatarLetter);
        }

        void bind(Contact contact, OnItemClickListener listener) {
            tvName.setText(contact.getName());
            tvPhone.setText(contact.getPhone());

            // Hiển thị ảnh nếu có, ngược lại hiện chữ cái đầu
            String base64 = contact.getPhoto();
            if (base64 != null && !base64.isEmpty()) {
                Bitmap bmp = ImageUtils.base64ToBitmap(base64);
                if (bmp != null) {
                    imgAvatar.setImageBitmap(bmp);
                    imgAvatar.setVisibility(View.VISIBLE);
                    tvAvatarLetter.setVisibility(View.GONE);
                } else {
                    showLetter(contact);
                }
            } else {
                showLetter(contact);
            }

            itemView.setOnClickListener(v -> listener.onItemClick(contact));
        }

        private void showLetter(Contact contact) {
            imgAvatar.setVisibility(View.GONE);
            tvAvatarLetter.setVisibility(View.VISIBLE);
            tvAvatarLetter.setText(ImageUtils.getAvatarLetter(contact.getName()));
            tvAvatarLetter.getBackground().setTint(ImageUtils.getAvatarColor(contact.getName()));
        }
    }
}

File: FresherAndroid/AndroidAdvanced/submit/ManhPT14_ReactProgr_Kotlinflow/app/src/main/java/com/example/contactappflow/ui/adapter/ContactAdapter.kt
package com.example.contactappflow.ui.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.contactappflow.data.local.Contact
import com.example.contactappflow.databinding.ItemContactBinding

class ContactAdapter(
    private val onEditClick: (Contact) -> Unit,
    private val onDeleteClick: (Contact) -> Unit
) : ListAdapter<Contact, ContactAdapter.ContactViewHolder>(DiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ContactViewHolder {
        val binding = ItemContactBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )
        return ContactViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ContactViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class ContactViewHolder(
        private val binding: ItemContactBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(contact: Contact) {
            binding.tvName.text = contact.name
            binding.tvPhone.text = contact.phone
            binding.btnEdit.setOnClickListener { onEditClick(contact) }
            binding.btnDelete.setOnClickListener { onDeleteClick(contact) }
        }
    }

    class DiffCallback : DiffUtil.ItemCallback<Contact>() {
        override fun areItemsTheSame(oldItem: Contact, newItem: Contact) =
            oldItem.id == newItem.id

        override fun areContentsTheSame(oldItem: Contact, newItem: Contact) =
            oldItem == newItem
    }
}